caas/kubernetes/cloud/credential.go

```diff
@@ -82,13 +82,131 @@ var SupportedCredentialSchemas = map[cloud.AuthType]cloud.CredentialSchema{
 	},
 }
 
+// LegacyCredentialsSchemas represents legacy credentials schemas that Juju used
+// to output but still need to be supported to maintain working Kubernetes
+// support. These types should be liberally allowed as input but not used as
+// new output from Juju. This change was introduced by tlm in juju 2.9
+var LegacyCredentialSchemas = map[cloud.AuthType]cloud.CredentialSchema{
+	cloud.OAuth2WithCertAuthType: {
+		{
+			Name: CredAttrClientCertificateData,
+			CredentialAttr: cloud.CredentialAttr{
+				Description: "the kubernetes certificate data",
+			},
+		},
+		{
+			Name: CredAttrClientKeyData,
+			CredentialAttr: cloud.CredentialAttr{
+				Description: "the kubernetes private key data",
+				Hidden:      true,
+			},
+		},
+		{
+			Name: CredAttrToken,
+			CredentialAttr: cloud.CredentialAttr{
+				Description: "the kubernetes token",
+				Hidden:      true,
+			},
+		},
+	},
+	cloud.CertificateAuthType: {
+		{
+			Name: CredAttrClientCertificateData,
+			CredentialAttr: cloud.CredentialAttr{
+				Description: "the kubernetes certificate data",
+			},
+		},
+		{
+			Name: CredAttrToken,
+			CredentialAttr: cloud.CredentialAttr{
+				Description: "the kubernetes service account bearer token",
+				Hidden:      true,
+			},
+		},
+		{
+			Name: RBACLabelKeyName,
+			CredentialAttr: cloud.CredentialAttr{
+				Optional:    true,
+				Description: "the unique ID key name of the rbac resources",
+			},
+		},
+	},
+}
+
+func migrateCertificateAuthType(cred *cloud.Credential) (cloud.Credential, error) {
+	attrs := cred.Attributes()
+	newAttrs := map[string]string{}
+
+	token, exists := attrs[CredAttrToken]
+	if !exists {
+		return cloud.Credential{}, errors.NotFoundf(
+			"certificate oauth token during migration, expect key %s",
+			CredAttrToken)
+	}
+
+	newAttrs[CredAttrToken] = token
+	if _, exists := attrs[RBACLabelKeyName]; exists {
+		newAttrs[RBACLabelKeyName] = attrs[RBACLabelKeyName]
+	}
+
+	return cloud.NewNamedCredential(
+		cred.Label,
+		cloud.OAuth2AuthType,
+		newAttrs,
+		false), nil
+}
+
+func MigrateLegacyCredential(cred *cloud.Credential) (cloud.Credential, error) {
+	switch cred.AuthType() {
+	case cloud.OAuth2WithCertAuthType:
+		return migrateOAuth2WithCertAuthType(cred)
+	case cloud.CertificateAuthType:
+		return migrateCertificateAuthType(cred)
+	default:
+		return cloud.Credential{}, errors.NotSupportedf(
+			"migration for auth type %s", cred.AuthType())
+	}
+}
```

Comment:
nit - can we move this up above the 2 unexported migrate functions
right now, it sits between them

Reply:
nah trying to keep things alphabetical ordered in the file which I have now fixed.